assets/write-asset-properties.js

```diff
@@ -0,0 +1,126 @@
+        APS_CLIENT_SECRET, 'data:read data:write');
+    const client = new TandemClient(() => {
+        return token;
+    });
+
+    // STEP 2 - get facility
+    const facilityId = FACILITY_URN;
+    const facility = await client.getFacility(facilityId);
+    // the map below contains property values for assets
+    // the key is asset id (stored in Device ID property - see above)
+    const assetPropertyMap = {
+        'P001' : {
+            'Controller Type': 'Centrifugal',
+            'Flow Rate': 2300,
+            'Frequency': 60,
+            'Temperature': 105,
+            'Working Pressure': 175
+        }
+    };
+
+    // STEP 3 - iterate through facility models
+    for (const link of facility.links) {
+        // STEP 4 - collect property names
+        const propertNames = new Set();
+        
+        for (const assetId in assetPropertyMap) {
+            const props = assetPropertyMap[assetId];
+
+            for (const propName in props) {
+                propertNames.add(propName);
+            }
+        }
+        // STEP 5 - create map between name and id
+        const schema = await client.getModelSchema(link.modelId);
+        const propertyMap = {};
+        let idProp; // use to store property wich is used to identify asset (i.e. Device ID by default)
+
+        for (const prop of schema.attributes) {
+            if (propertNames.has(prop.name)) {
+                propertyMap[prop.name] = prop;
+            }
+            if (prop.name === ASSET_ID_PROPERTY) {
+                idProp = prop;
+            }            
+        }
+        if (!idProp) {
+            continue;
+        }
+        // STEP 6 - iterate through assets and collect changes
+        const assets = await client.getTaggedAssets(link.modelId);
+        const keys = [];
+        const mutations = [];
+        
+        for (const asset of assets) {
+            const deviceId = asset[idProp.id];
+
+            if (!deviceId) {
+                continue;
+            }
+            console.log(`${asset[QC.Name]}:${deviceId}`);
+            // check for new property data from map above
+            const propertyData = assetPropertyMap[deviceId];
+
+            if (!propertyData) {
+                continue;
+            }
+            for (const prop in propertyData) {
+                const newValue = propertyData[prop];
+                const propDef = propertyMap[prop];
+
+                if (!propDef) {
+                    continue;
+                }
+                const currentValue = asset[propDef.id];
+                
+                // we apply change only if there is difference
+                if (currentValue === newValue) {
+                    continue;
+                }
+                console.log(`  ${propDef.name}:${newValue}`);
+                keys.push(asset[QC.Key]);
+                mutations.push([
+                    MutateActions.Insert,
+                    propDef.fam,
```

Comment:
wouldn't it always be the same family in the context of this example? I mean `ColumnFamilies.DtProperties`? That's practically the common case for writing properties back and I think we need to be clear about that.

Reply:
Agree, added check and comment into code.